### PR TITLE
Update remote app builder pattern

### PR DESCRIPTION
**Pull request by twsouthwick**
**PR Title**
Update builder pattern for remote to clearly separate client and server scenarios

**PR Description**
This change makes a few API changes we've discussed around being clear around remote app options and extension methods. This introduces a new builder API (ISystemWebAdapterRemoteAppBuilder) that allows for a nested configuration pattern similar to the following:

**ASP.NET Core:**

```csharp
builder.Services.AddSystemWebAdapters()
  .AddRemoteAppClient(remote => remote
    .Configure(options =>
    {
      options.RemoteAppUrl = ...
      options.ApiKey = ...
    }
    .AddAuthentication(true)
    .AddSession());
```

**ASP.NET Framework:**

```csharp
SystemWebAdapterConfiguration.AddSystemWebAdapters(this)
    .AddRemoteAppServer(remote => remote
        .Configure(options => options.ApiKey = ... )
        .AddAuthentication()
        .AddSession());
```

This change also separates out the options for Client/Server to be separate classes

Part of #112

### CODE REVIEW

File: src/Microsoft.AspNetCore.SystemWebAdapters/Microsoft.AspNetCore.SystemWebAdapters.csproj [hunk 1]
```diff
@@ -66,6 +66,10 @@
     <PackageReference Include="Microsoft.Extensions.Options" Version="6.0.0" />
   </ItemGroup>
 
+  <ItemGroup>
+    <SharedFiles Remove="Adapters\ISystemWebAdapterRemoteAppBuilder.Shared.cs" />
```
Comment (mjrousos) [inline src/Microsoft.AspNetCore.SystemWebAdapters/Microsoft.AspNetCore.SystemWebAdapters.csproj:70]: Why is this removed?

File: src/Microsoft.AspNetCore.SystemWebAdapters/Adapters/ISystemWebAdapterRemoteAppBuilder.Shared.cs [hunk 2]
```diff
@@ -0,0 +1,11 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using Microsoft.Extensions.DependencyInjection;
+
+namespace Microsoft.AspNetCore.SystemWebAdapters;
+
+public interface ISystemWebAdapterRemoteAppBuilder
```
Comment (mjrousos) [inline src/Microsoft.AspNetCore.SystemWebAdapters/Adapters/ISystemWebAdapterRemoteAppBuilder.Shared.cs:8]: What if we had two remote app builder interfaces - a client version and a server version. They could have the same shape (in fact, they could both just derive from `ISystemWebAdapterRemoteAppBuilder` with no changes) but then all of the methods that `ISystemWebAdapterRemoteAppBuilder` wouldn't need client or server in their names. You could just have `AddRemoteAppAuthentication` and have the client one extend `ISystemWebAdapterRemoteAppClientBuilder` while the server one extends `ISystemWebAdapterRemoteAppServerBuilder`.

**Review by mjrousos — CHANGES_REQUESTED** [review]
Looks good. As I looked at the changes to samples, though, it occurred to me that we might be able to simplify the API a little by having `AddRemoteAppServer` and `AddRemoteAppClient` configure different interfaces (`ISystemWebAdapterRemoteAppClientBuilder` and `ISystemWebAdapterRemoteAppServerBuilder`). Then the extension methods that work on those builders wouldn't all need the "client" and "server" designations (since they would extend different interfaces) and overall usage would simplify a little.

File: src/Microsoft.AspNetCore.SystemWebAdapters/Microsoft.AspNetCore.SystemWebAdapters.csproj (hunk 1 above)
Comment (twsouthwick) [reply src/Microsoft.AspNetCore.SystemWebAdapters/Microsoft.AspNetCore.SystemWebAdapters.csproj:70]: oh VS being helpful.... I'll remove this

File: src/Microsoft.AspNetCore.SystemWebAdapters/Adapters/ISystemWebAdapterRemoteAppBuilder.Shared.cs (hunk 2 above)
Comment (twsouthwick) [reply src/Microsoft.AspNetCore.SystemWebAdapters/Adapters/ISystemWebAdapterRemoteAppBuilder.Shared.cs:8]: those type names don't show up anywhere, though, so I'm not sure how helpful that would be. I was wondering if we even needed the "client/server" after the initial one, i.e.

**ASP.NET Core:**

```csharp
builder.Services.AddSystemWebAdapters()
  .AddRemoteAppClient(remote => remote
    .Configure(options =>
    {
      options.RemoteAppUrl = ...
      options.ApiKey = ...
    }
    .AddAuthentication(true)
    .AddSession());
```

**ASP.NET Framework:**

```csharp
SystemWebAdapterConfiguration.AddSystemWebAdapters(this)
    .AddRemoteAppServer(remote => remote
        .Configure(options => options.ApiKey = ... )
        .AddAuthentication()
        .AddSession());
```

It's already being done as part of `AddRemoteApp[Server|Client]` so this seems like it would be fairly clear.

Comment (mjrousos) [reply src/Microsoft.AspNetCore.SystemWebAdapters/Adapters/ISystemWebAdapterRemoteAppBuilder.Shared.cs:8]: Even if the user doesn't see the type names, I think it would help because it would allow us to have the API you're describing but only the correct `Configure` or `AddAuthentication` would be usable (and suggested by Intellisense) because they would extend different types. As it is now, wouldn't the user have to choose between to `Configure` methods? That's not a great experience which is why we were going to name them Client and Server. I think if they extend different types, though, then we can get away with them both just being called 'Configure'

Comment (twsouthwick) [reply src/Microsoft.AspNetCore.SystemWebAdapters/Adapters/ISystemWebAdapterRemoteAppBuilder.Shared.cs:8]: Oh - the client stuff is only available on the core side while the framework is only available on the framework. But I see your point. I'll try that out

Comment (mjrousos) [reply src/Microsoft.AspNetCore.SystemWebAdapters/Adapters/ISystemWebAdapterRemoteAppBuilder.Shared.cs:8]: Ah, good point. I forgot that even though `ISystemWebAdapterRemoteApp` is shared, the extension methods aren't! In that case, it doesn't make as big of a difference. I still think it would be nice to have client and server versions of the builder so that it's clearer which the extension methods apply to without having to give them cumbersome names.

Comment (twsouthwick) [reply src/Microsoft.AspNetCore.SystemWebAdapters/Adapters/ISystemWebAdapterRemoteAppBuilder.Shared.cs]: I made a separate builder for both

File: src/Microsoft.AspNetCore.SystemWebAdapters/Adapters/RemoteAppServerExtensions.Framework.cs [hunk 3]
```diff
@@ -0,0 +1,63 @@
...
+
+namespace System.Web;
+
+public static class RemoteAppServerExtensions
+{
+    /// <summary>
+    /// Add configuration for connecting to a remote app for System.Web extensions that require a remote
+    /// ASP.NET app such as remote app authentication or remote app session sharing.
+    /// </summary>
+    public static ISystemWebAdapterBuilder AddRemoteAppServer(this ISystemWebAdapterBuilder builder, Action<ISystemWebAdapterRemoteServerAppBuilder> configure)
```
Comment (Tratcher) [inline src/Microsoft.AspNetCore.SystemWebAdapters/Adapters/RemoteAppServerExtensions.Framework.cs:15]: Having both a builder callback and a builder return value on the same API is an odd mixture of patterns. Though I can also appreciate that returning ISystemWebAdapterRemoteServerAppBuilder here could have also been confusing and would break the ISystemWebAdapterBuilder chaining.

**Review by Tratcher — APPROVED** [review]
We'll cover this in detail in the next API review.

**Review by mjrousos — APPROVED** [review]
Looks good. We should update docs to reflect the changes, too. If you want to do that in a separate PR, though, this part looks good.

**Comment by twsouthwick** [comment]
:) I was in the process of updating the docs so I wouldn't forget

File: src/Microsoft.AspNetCore.SystemWebAdapters/Adapters/RemoteAppServerExtensions.Framework.cs (hunk 3 above)
Comment (twsouthwick) [reply src/Microsoft.AspNetCore.SystemWebAdapters/Adapters/RemoteAppServerExtensions.Framework.cs:15]: yeah - I had the same thought, but seemed to make more sense than having a new builder. We can discuss it at the API review
